### DISCUSSION TITLE
.NET 8.0 Preview 1 release - .NET 8.0.0-preview.1 and SDK 8.0.100-preview.1

### DISCUSSION THREAD

**Discussion by MichaelSimons** [dotnet/source-build#3263] [Announcements]
Please use the [v8.0.100-preview.1](https://github.com/dotnet/dotnet/releases/tag/v8.0.100-preview.1) tag to source-build .NET version 8.0.0-preview.1 / 8.0.100-preview.1.

* Release Notes: https://github.com/dotnet/core/blob/main/release-notes/8.0/preview/8.0.0-preview.1.md
* Tag URL: https://github.com/dotnet/dotnet/releases/tag/v8.0.100-preview.1

## Virtual Mono Repository (VMR)

.NET is now buildable on Linux directly from the [dotnet/dotnet](https://github.com/dotnet/dotnet) repository. We're calling this new repository a [Virtual Mono Repository (VMR)](https://github.com/dotnet/arcade/issues/11584). It has the benefits of a true [monorepo](https://en.wikipedia.org/wiki/Monorepo) but is a [regularly-updated projection](https://github.com/dotnet/dotnet/commits/main) of the many existing repos that contributors work in (more efficiently) every day. We see this new approach as being a significant step forward in approachability for building .NET as a whole product from source.

See [build instructions](https://github.com/dotnet/dotnet#building) to build the VMR on your own machine.

## Prebuilts

This release requires the following prebuilts.  These prebuilts are retrieved by `prep.sh` in order to support building in an offline environment.

* Microsoft.AspNetCore.App.Ref.6.0.13
* Microsoft.AspNetCore.App.Ref.7.0.2
* Microsoft.AspNetCore.App.Runtime.osx-x64.7.0.2
* Microsoft.AspNetCore.App.Runtime.win-x64.7.0.2
* Microsoft.AspNetCore.App.Runtime.win-x86.7.0.2
* Microsoft.NET.ILLink.Tasks.7.0.100-1.23062.2
* Microsoft.NETCore.App.Host.linux-x64.7.0.2
* Microsoft.NETCore.App.Host.osx-x64.7.0.2
* Microsoft.NETCore.App.Host.win-x64.7.0.2
* Microsoft.NETCore.App.Host.win-x86.7.0.2
* Microsoft.NETCore.App.Ref.6.0.13
* Microsoft.NETCore.App.Ref.7.0.2
* Microsoft.NETCore.App.Runtime.osx-x64.7.0.2
* Microsoft.NETCore.App.Runtime.win-x64.7.0.2
* Microsoft.NETCore.App.Runtime.win-x86.7.0.2
* Microsoft.NETFramework.ReferenceAssemblies.1.0.3
* Microsoft.NETFramework.ReferenceAssemblies.net20.1.0.3
* Microsoft.NETFramework.ReferenceAssemblies.net45.1.0.3
* Microsoft.NETFramework.ReferenceAssemblies.net452.1.0.3
* Microsoft.NETFramework.ReferenceAssemblies.net462.1.0.3
* Microsoft.NETFramework.ReferenceAssemblies.net472.1.0.3
* Microsoft.NETFramework.ReferenceAssemblies.net48.1.0.3

## Bootstrapping

Each major release of .NET requires bootstrapping. Because of the previously mentioned prebuilts, the stage 2 build (e.g. rebuilding) requires the following RID-specific prebuilt. This needs to be retrieved from your 7.0.2 source-built artifacts.  If you haven't source-built 7.0.2 you won't be able to bootstrap 8.0 preview 1.  This prebuilt is not required when building stage 1 (e.g. building with the Microsoft provided SDK).  Place this package into the directory specified by the `build.sh --with-packages` option.

* Microsoft.NETCore.App.Host.`<RID>`.7.0.2

@dotnet/distro-maintainers

```json
{
  "Channel": "8.0",
  "Runtime": "8.0.0-preview.1",
  "Sdk": "8.0.100-preview.1",
  "DotnetTag": "v8.0.100-preview.1"
}
```

**Comment by ayakael** [comment]
Very much appreciate the list of required prebuilts! 

I'm glad to report that an aport for dotnet8 on `alpine.3.18-x64` is working, and a final package is being built on my [Gitlab instance](https://lab.ilot.io/ayakael/user-aports) with artifacts automatically pushed [here](https://lab.ilot.io/ayakael/repo-apk). Only GA of dotnet8 will be pushed to the official repo, but I hope that this will make sure broken bits are found along the way.

In somewhat related news, I'm working on automating maintenance for dotnet on Alpine. I've so far automated tarballs building for [dotnet6](https://lab.ilot.io/dotnet/installer/-/releases), [dotnet7](https://lab.ilot.io/dotnet/installer/-/releases) and [dotnet8](https://lab.ilot.io/dotnet/dotnet/-/releases). Fellow package maintainers are welcome to use these, although as I run this server at home I cannot ensure 100% availability. Next step is automating bootstrap crossbuilding from `linux-musl-x64` to `linux-musl-s390x` / `linux-musl-ppc64le` as to facilitate testing for those platforms as net8 development moves along. 

Very happy with the VMR changes! Thanks for all the work.

**Comment by tmds** [comment]
> I've so far automated tarballs building for [dotnet6](https://lab.ilot.io/dotnet/installer/-/releases), [dotnet7](https://lab.ilot.io/dotnet/installer/-/releases) and [dotnet8](https://lab.ilot.io/dotnet/dotnet/-/releases).

Nice! There was an issue for source-build to provide tarballs: https://github.com/dotnet/source-build/issues/298.
If there is still a common need, it would be nice if the upstream project could provide these.

Are you also creating a tarball somewhere that contains the prebuilts?

**Reply by ayakael** [reply]
> > I've so far automated tarballs building for [dotnet6](https://lab.ilot.io/dotnet/installer/-/releases), [dotnet7](https://lab.ilot.io/dotnet/installer/-/releases) and [dotnet8](https://lab.ilot.io/dotnet/dotnet/-/releases).
> 
> Nice! There was an issue for source-build to provide tarballs: #298. If there is still a common need, it would be nice if the upstream project could provide these.
> 
> Are you also creating a tarball somewhere that contains the prebuilts?

Indeed, I was tracking that till I got my GitLab instance going, and decided that it was worth implementing myself. I would love for upstream to provide these instead.

For now I have no plans to host prebuilt tarballs as for our use-case it is not necesary; `dotnetx-stage0` packages platform-agnostic Microsoft prebuilts and builds the rest so that `dotnetx-build` can ingest it. That said, I am exploring GitLab's nuget repo function.

**Reply by tmds** [reply]
For dotnet8, if your tarball is exactly the same as the `dotnet/dotnet` repo, you can probably directly use the .tar.gz source code link from the releases page: https://github.com/dotnet/dotnet/releases/tag/v8.0.100-preview.1.

**Reply by ayakael** [reply]
Unfortunately, those files are missing `.git/HEAD` and `.git/config`. `eng/pack-source.sh` adds those when generating the tarball. It be nice if that information was pushed to `prereqs/git-info` as to allow directly using the .tar.gz in the releases page.

**Reply by premun** [reply]
You can use [this script](https://github.com/dotnet/dotnet/blob/main/eng/pack-sources.sh) to export the sources with the HEAD/config files included. We have coverage in our official CI, where for some specific distro builds that we're testing, we use this script first and then build from the export. So the build should work.

We should mention this in the main README. As this has been the first release there, I agree it might be a good idea to include this export instead of the default sources that GitHub appends to each release. We will look into that.

**Reply by MichaelSimons** [reply]
> Are you also creating a tarball somewhere that contains the prebuilts?

Yes, `prep.sh` will retrieve a prebuilt tarball and the previous source-build artifacts necessary to build with.  After running the script you will find these at ./prereqs/packages/archive.

**Reply by tmds** [reply]
> It be nice if that information was pushed to prereqs/git-info as to allow directly using the .tar.gz in the releases page.

+1

**Reply by MichaelSimons** [reply]
I logged https://github.com/dotnet/source-build/issues/3265 to track providing a buildable source-tarball.

**Comment by omajid** [comment]
I have created a work-in-progress package for Fedora-family distros based on this 8.0 Preview 1 release.

Sources: https://pagure.io/dotnet-sig/dotnet8.0
Binaries: https://copr.fedorainfracloud.org/coprs/g/dotnet-sig/dotnet-preview/

Binaries for 8.0 Preview 1 are available for Fedora (36, 37, 38, Rawhide), CentOS Stream (8 and 9), and RHEL (8 and 9) on x86_64 (aka x64) and aarch64 (aka arm64). The copr build system doesn't let me build RHEL8+aarch64, so that's not available.

**Reply by tmds** [reply]
This is awesome!

I think last time we did a preview release on Fedora, it was for one of the later previews of .NET 3 in the summer of 2019.

Being able to build a preview 1 in tandem with Microsoft is a major milestone.
Great work @dotnet/source-build-contrib for making this possible!

@omajid, thank you for providing the builds for the Fedora-based distros!
